### DISCUSSION TITLE
Diffraction close to the rotation axis

### DISCUSSION THREAD

**Discussion by Marcraven** [FABLE-3DXRD/xrd_simulator#18] [General]
The missing diffraction spots around the rotation axis region become quite evident when simulating many many tetrahedra.

I think the reason behind this is that the beam is perfectly monochromatic and also perfectly parallel. That way, it is  very unlikely to get diffraction close to the rotation direction, since the G vector should lay exactly or very close to the vertical plane for a diffraction spot to lay there, and rotation around its axis would not help.

Is there a discussion about how to improve on this?

![Image 2024-11-07 at 103033](https://github.com/user-attachments/assets/f557328d-85e0-475c-a922-6fde358a8a5c)
![Image 2024-11-07 at 102832](https://github.com/user-attachments/assets/7b52f29f-8edb-4e7a-8b14-2326d3de7b68)

**Comment by AxelHenningsson** [comment]
Hi @Marcraven !

Indeed, it is a feature of monochromatic beams. This is also what is typically seen in synchrotron diffraction experiments when the beam is well monochromated.

To change this amounts to implementing something like an energy bandwith for the incoming beam. Letting wavelength be variable in the the Laue equations fundamentaly changes the algeabraic solution procedure which `xrd_simulator` is built upon.[ As described in this publication.](https://journals.iucr.org/j/issues/2023/01/00/nb5332/index.html)

I belive there is no trivial "quick fix" to implement energy bandwidth in terms of changing the core diffraction code (if you can come up with one it would be cool)

For some other projects I have seen that people typically iterate over a fixed grid of wavelengths and solve the Laue equations at each point to then add the scattering contribution for each energy (weighted by the spectrum curve). This would amount to computing a lot more scattering units in `xrd_simulator` and is equivalent of running `xrd_simulator` multiple times at different wavelengths weighting the resulting diffraction patterns together. That is very computationally expensive though, depending on the setup.

Cheers
Axel

**Reply by Marcraven** [reply]
Hi @AxelHenningsson ,

Yes that is what I thought...

I implemented a Laue simulator a while ago that might be worth implementing here. Then the user could chose to go monochromatic or laue.

Don't you think that the lack of beam divergence is not also causing this missing angle?

However, and before jumping into this laue, I have a branch which I think would be worth merging, but it changes huge parts of the code... _again..._

For what is worth, I am now simulating 500.000 tetrahedra in 2 seconds with GPU and in 12 seconds or so in CPU, after moving all the code to PyTorch.

I also made the diffraction and rendering independent from each other, so one can diffract from the crystal and then render as many detectors in arbitrary positions from those diffraction vectors.

I think we should organize a meeting eventually to sort out these changes.

Cheers,
